### DISCUSSION TITLE
Purpose of headless module

### DISCUSSION THREAD

**Discussion by bevans2000** [mars-sim/mars-sim#312] [Ideas]
I was wondering about the headless mode. It still uses a GUI Swing components but I would expect it to be totally text based. The headless module could go text only and then be provided as Docker image. This would allow the simulation to be run easily in a cloud container to allow very long simulation runs. A remote version of the Console could connect to the remote executable.
Something like  a SSH connection would be good so existing SSH clients could be used like Putty or Terminus.

**Comment by Urwumpe** [comment]
I agree there - AFAIR, we have some lot of old code and assumptions about Swing being used in the background in our core module. A real headless mode should be better. 

A later WS version could also profit from experiences with a real headless mode and Docker images - we could maybe reuse a lot of tools and tests then.

**Comment by bevans2000** [comment]
I have a very basic Docker version running which I can connect to the Console via a SSH client (I'm using Putty). At the moment there is a single "admin" user that is equivalent to the UI Console access. One thought I have had on this was how it goes with the Commander logic. 
The Commander logic at the moment considers there can only be a single Commander but if it runs as a service then there could be multiple Commanders concurrently; one per Settlement. So I wondered if the singleton Commander logic should be ripped out and replaced with the ability to connect to the existing Commander Person of a Settlement. 
Furthermore, the new Console Chat framework would support an extension such that the commands and output come be done over a WebService as well. So this may eventually allow us to control a shared simulation via a web service/JSON API.

**Comment by mokun** [comment]
> I was wondering about the headless mode. It still uses a GUI Swing components but I would expect it to be totally text based. 

Sure. I'm more than willing to see a true headless mode. 

Currently, it requires a package called `text-io` which in terms relies on Swing. 

I imagine in future, we can have a machine running the server version of mars-sim, while allowing multiple clients connecting to the server.

Each client is responsible for one settlement. At some point, it can even start remote stations/settlement and form a government of its own

OR if a player is a CEO running a corporation, he can be in charge of multiple locations on Mars.

Together, it comprises one massive Mars simulation.

**Comment by bevans2000** [comment]
At the moment the GameMode only allows one Commander per simulation. I think we need to drop it and allow one Commander per Settlement. However I do not understand how the CommanderProfile is expected to work in the UI  mode. I would like to remove the Commander Profile but replace it the ability to take control of Settlement by controlling the existing elected person with the Commander role.
I'll be honest I missed the Commander capability when I reworked the Console, so what should a Commander do?

**Reply by mokun** [reply]
> At the moment the GameMode only allows one Commander per simulation. I think we need to drop it and allow one Commander per Settlement.

Each settlement always have at least one commander elected. 

If a settlement has 8+ settlers, it will elect a sub-commander as well.

May be you mean in the `Command Mode`, whereas a player can customize the commander's various attributes.

This is in contrast with `Sandbox Mode`, in which a player cannot designate who the commander will be. A commander is elected based on one's leadership and other attributes.

**Reply by bevans2000** [reply]
OK, my mistake. So in `Command Mode` does the user have any extra capabilities via the Console or UI? The fact it is a singleton creates a blocker for multiple users using a single shared simulation service. 
Maybe one way to address this is to allow remote user the ability to control a Settlement as a proxy-Commander (only one user per Settlement). Then they could change the attributes of the elected Commander as in the `Command Mode`. Also give them the ability to create new Missions via the remote connection. If a remote user is assigned to a Settlement then that would also limit what Units they can see; only those associated to the Settlement. Probably would also need the ability to see the summary stats of other Settlements to help decide on Missions. For Settlements with the same Sponsor; more details could be displayed.

**Reply by mokun** [reply]
> So in `Command Mode` does the user have any extra capabilities via the Console or UI? 

Yes. In `Command Mode`, players are given an extra window panel called [Commander Tool](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-ui/src/main/java/org/mars_sim/msp/ui/swing/tool/commander/CommanderWindow.java). 

However, out of the 7 tabs inside, we only put Logistic and Mission into use. More could have been done in near future.

```
public static final String AGRICULTURE_TAB = "Agriculture";
	public static final String ENGINEERING_TAB = "Engineering";
	public static final String LOGISTIC_TAB = "Logistic";
	public static final String MISSION_TAB = " Mission";
	public static final String RESOURCE_TAB = "Resource";
	public static final String SAFETY_TAB = "Safety";
	public static final String SCIENCE_TAB = "Science";
```

**Reply by bevans2000** [reply]
Right I see. I think we could create a new `Commander` class that acts as a Facade to the various lower level classes. This could be used by both the Console and UI to centralise the logic. The new class would be constructed from a Settlement. This remove the need for the singleton GameMode class.

**Reply by mokun** [reply]
> ve the need for the singleton GameMode class.



> Right I see. I think we could create a new `Commander` class that acts as a Facade to the various lower level classes.

Sure !

> This could be used by both the Console and UI to centralise the logic. The new class would be constructed from a Settlement. This remove the need for the singleton GameMode class.

Yes it doesn't have to be a singleton.

**Reply by Urwumpe** [reply]
> Right I see. I think we could create a new `Commander` class that acts as a Facade to the various lower level classes. This could be used by both the Console and UI to centralise the logic. The new class would be constructed from a Settlement. This remove the need for the singleton GameMode class.

I agree there, but maybe its smarter to construct this in its own Factory class, instead of delegating this to Settlement. A commander should be assigned to a scope, right now its the Settlement, so its fine to assume a 1:1-relation between commander and Settlement.

Also, the new Commander class should then know as little as needed about the implementation of the actions that it triggers. Ideally, it would be just like making one chosen Action the current action of the Settler, without the Commander object even knowing if the Settler is currently capable of it, how the action is implemented and which effects the action has on the Settler entity. 

Otherwise, we will have too many implicit assumptions in our code about things that will change in the future... never a wise thing.

**Comment by mokun** [comment]
@bevans2000 ,

For the Swing Edition, we should probably switch to a graphical interface for the player to choose the game mode and whether to start the Site Editor / Crew Editor.

Currently, both the Console and Swing Edition require the player to go through the MAIN MENU in the text-io console in order to start a brand new mars-sim.

**Comment by mokun** [comment]
@bevans2000 ,

Shall we use spring boot for the server/client version of mars-sim, instead of using AWS ?

This way, I can easily run mars-sim across machines on an internal network at home.

**Reply by bevans2000** [reply]
The use of spring boot & AWS are separate things and shouldn't cause a problem. If we create a Docker image of the headless mars-sim then this can be run on AWS, Azure, local Docker installation or Kubernetes etc etc. It's just a means to conveniently packages services for easy deployment. The headless can still be run as a stand-alone Java app on the command line as well.
Whether we use Spring boot or not has no bearing on the Docker image. Yes Spring boot does have some nice integration with Docker runtime but it's not important for mars-sim really.
II should have a PullRequest ready within a day or so.

**Comment by bevans2000** [comment]
There is a Pull Request #316  with an implementation of a new headless mode supporting SSH connections for remote clients.
I choose SSH as it's an easy, well support & secure protocol. There are lots of client tools out there on Windows, Android & Linux. 
I'm not saying it right for the long term but certainly a good stop gap in the medium term. Ideally I would like to put a web service on top of the simulation; maybe mapping onto the existing Command framework.

**Reply by mokun** [reply]
@bevans2000 ,

Can you jot down some notes as to how to set up SSH connection in Windows ? 

Let's pick one tool that you use in Windows and give players a concrete example on how to set it up.

Appreciate !

**Reply by bevans2000** [reply]
I can write a page on the wiki.

**Comment by bevans2000** [comment]
https://github.com/mars-sim/mars-sim/wiki/Remote-Console-Connection

**Reply by mokun** [reply]
> https://github.com/mars-sim/mars-sim/wiki/Remote-Console-Connection

Thanks @bevans2000 

For now, it's being placed under "A. Introduction". 

I guess it's still NOT stable enough to do an official release of 3.1.2 until we find the airlock walking bug at least, right ?

**Reply by bevans2000** [reply]
Yes I feel the airlock bug needs fixing. Whilst looking into #314 I have found that some Tasks are returning more remaining time than there was originally. This is dangerous because the  I d could go I to a loop never returning.

**Comment by bevans2000** [comment]
I have added a new Github Flow to automatically build a Docker image when there is a push to the master branch. This is gets pushed to the GitHub package repository and can be used directly from there.
https://github.com/orgs/mars-sim/packages?repo_name=mars-sim

**Reply by mokun** [reply]
> This is gets pushed to the GitHub package repository and can be used directly from there.

Great ! So is there a way to test run the docker image remotely online somewhere, without having to set up my own local machine at home ?

**Reply by bevans2000** [reply]
I run it on AWS ECS because I use AWS a lot. There a few cloud providers who host Docker; you can just google. 
However I have found out that the GitHub Packages doesn't seem to support anonymous pull requests. The github packages can't be used in a docker pull request without authentication.
So it might be better to turn the Docker building over to DockerHub (https://docs.docker.com/docker-hub/builds/) and create a mars-sim Docker Organisation to hold the builds. Certainly most people expect images to be at DockerHub. 
Not sure if this is worth the effort yet. Proved that the Docker route does work which is a start.

**Comment by bevans2000** [comment]
If we want to pursue the Docker option then we probably need to create an Organisation on DockerHub and get that to do the builds for us. This would need one of us with GitHub admin right to the project to create a webhook to trigger the Docker builds. This would make the Docker image more readily available as all hosted Docker provides can use DockerHub.
Probably need a little work on the Dockerfile to allow outside control on whether to create a new simulation or re-use a saved one.
Docker is a great way of delivering services for deployment so if the project is to move out of the local laptop model then it's a good route to take.
